### PR TITLE
Support nocheck prefix

### LINKED ISSUE #72695: ILDAsm and ILAsm don't handle the `no.` prefix
### Description

ILAsm currently errors out when encountering IL containing the `no.` prefix, while ILDAsm disassemblers it into output that isn't valid.

### Reproduction Steps

For ILDAsm: Create an assembly containing `no.` with some value (possible with Cecil) and try to disassemble it.

For ILAsm: Try to assemble:
```cil
ldarg.0
ldc.i4.0
no. 0x06
ldelem.u1
ret
```

### Expected behavior

Both are correctly handled.

### Actual behavior

ILDasm gives for `no.` with 0x06:
```cil
ldarg.0
ldc.i4.0
unused
ldloc.0
delem.u1
ret
```

ILAsm errors out with `error : syntax error at token 'no' in:     IL_0002:  no. 0x02`.

### Regression?

VS ILDasm seems to have the same behaviour.

### Known Workarounds

Using Cecil to create the instruction.

### Configuration

.Net 7-preview 6 ILDasm & ILAsm

### Other information

Fixing this is required for creation of tests for eventual runtime support for the instructions with is tracked by #10112.

### PR DESCRIPTION
**Pull request by john-h-k**
Another shot at #35491

Supports the `no.` prefix as a nop

I have a separate branch implementing it but I think by splitting it up it will be a bit more simple.

Fixes #72695

### CODE REVIEW

**Comment by john-h-k** [comment]
Fixes #72695

**Comment by john-h-k** [comment]
This is currently blocked as someone with the Microsoft internal YACC tool needs to regenerate asmparse.cpp

**Comment by JulieLeeMSFT** [comment]
Thanks @john-h-k, @jakobbotsch will help you on this.

> This is currently blocked as someone with the Microsoft internal YACC tool needs to regenerate asmparse.cpp

**Comment by jakobbotsch** [comment]
@john-h-k I've regenerated the parser with a small fix in the grammar. Can you test whether it works as expected? We should also verify that roundtripping with ildasm works correctly.

File: src/coreclr/jit/fgbasic.cpp [hunk 1]
```diff
@@ -2084,6 +2084,19 @@ void Compiler::fgFindJumpTargets(const BYTE* codeAddr, IL_OFFSET codeSize, Fixed
                 goto OBSERVE_OPCODE;
             }
 
+            case CEE_NOCHECK:
+            {
+                noway_assert(sz == 0);
+
+                int flags = getU1LittleEndian(codeAddr);
+
+                prefixFlags |= (flags <<= 6);
```
Comment (jakobbotsch) [inline src/coreclr/jit/fgbasic.cpp:2093]: Can we just have the three flag checks and avoid having this cross-enum dependency?

File: src/coreclr/jit/fgbasic.cpp [hunk 2]
```diff
@@ -2084,6 +2084,19 @@ void Compiler::fgFindJumpTargets(const BYTE* codeAddr, IL_OFFSET codeSize, Fixed
...
+
+                impValidateCheckElisionOpcode(codeAddr, codeEndp, true);
```
Comment (jakobbotsch) [inline src/coreclr/jit/fgbasic.cpp:2095]: Doesn't this pass `true` for the flags?

**Comment by JulieLeeMSFT** [comment]
@john-h-k, checking back if you will work on this PR soon. If it is not handled for too long, this PR will be closed automatically.

**Comment by john-h-k** [comment]
@JulieLeeMSFT Yes this is still WIP, have addressed @jakobbotsch 's comments

File: src/coreclr/jit/importer.cpp [hunk 3]
```diff
@@ -8753,6 +8787,36 @@ void Compiler::impImportBlockCode(BasicBlock* block)
...
+                    int flags = getU1LittleEndian(codeAddr);
+
+                    JITDUMP(" %u", flags);
+
+                    ++codeAddr;
+
+                    // PREFIX_NO_TYPECHECK = 0x000000040 which is 0x1 (the value of typecheck flag) << 6
+                    // The 2 subsequent flags are just left shift of 1, the same as the nullcheck/rangecheck IL flags, so we just have to
+                    // left shift by 6 to transform
+                    flags <<= 6;
```
Comment (jakobbotsch) [inline src/coreclr/jit/importer.cpp:8805]: This should be changed in the same way as `fgFindJumpTargets`.

File: src/coreclr/jit/importer.cpp [hunk 4]
```diff
@@ -8753,6 +8787,36 @@ void Compiler::impImportBlockCode(BasicBlock* block)
...
+                    if ((flags & (~PREFIX_NO_TYPERANGENULLCHECK)) != 0)
+                    {
+                        BADCODE("no. followed by invalid flags");
+                    }
+
+                    impValidateCheckElisionOpcode(codeAddr, codeEndp, flags);
+
+
+
```
Comment (jakobbotsch) [inline src/coreclr/jit/importer.cpp:8814]:
```suggestion

```

File: src/coreclr/vm/readytorunstandalonemethodmetadata.cpp [hunk 5]
```diff
@@ -93,9 +93,6 @@ class ReadyToRunStandaloneMethodMetadataHelper
                 case CEE_STLOC_S:
                 case CEE_LDC_I4_S:
                 case CEE_UNALIGNED:
-                case CEE_UNUSED69: // This is the no. prefix that is partially defined in Partition III.
-                    SkipIL(1);
-                    break;
```
Comment (jakobbotsch) [inline src/coreclr/vm/readytorunstandalonemethodmetadata.cpp:98]: Shouldn't we still skip 1 byte due to the flags?

**Comment by BruceForstall** [comment]
@john-h-k did you see the comments from @jakobbotsch ?

**Comment by john-h-k** [comment]
Have seen the comments and will address ASAP

**Comment by TIHan** [comment]
@john-h-k following up - what is your status with this PR?

**Pull request closed without merging**